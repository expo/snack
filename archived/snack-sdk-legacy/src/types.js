/**
 * @flow
 */

import type { SDKVersion } from './configs/sdkVersions';

type RequiredSnackFileAttributes = {
  contents: string,
  type: 'ASSET' | 'CODE',
};

type ExpoRequiredSnackFiles = {
  'app.js': RequiredSnackFileAttributes,
};

export type ExpoSnackFiles = {
  ...$Exact<ExpoRequiredSnackFiles>,
  ...{ [string]: RequiredSnackFileAttributes },
};

export type ExpoWebPlayer = {
  subscribe: () => void,
  unsubscribe: () => void,
  publish: (message: any) => void,
  listen: (listener: (message: any) => void) => void,
};

export type ExpoSnackSessionArguments = {
  files: ExpoSnackFiles,
  sdkVersion?: SDKVersion,
  verbose?: boolean,
  sessionId?: string, // Will be randomly generated if not provided
  host?: string,
  sessionSecret: string,
  snackId?: string,
  name?: string,
  description?: string,
  dependencies?: any, // TODO: more specific
  authorizationToken?: string,
  disableDevSession?: boolean,
  user: { idToken?: ?string, sessionSecret?: ?string },
  deviceId: ?string,
  player?: ExpoWebPlayer,
  isFocused?: boolean,
};

export type ExpoSubscription = {
  remove: () => void,
};

// Called with empty array if errors have been resolved
export type ExpoErrorListener = (errors: Array<ExpoError>) => void;

export type ExpoLogListener = (log: ExpoDeviceLog) => void;

export type ExpoPresenceListener = (event: ExpoPresenceEvent) => void;

export type ExpoStateListener = (event: ExpoStateEvent) => void;

export type ExpoDependencyErrorListener = (message: string) => void;

export type ExpoSendBeaconCloseRequest = {
  url: string,
  data: Blob,
};

export type ExpoSendBeaconCloseRequestListener = (closeRequest: ExpoSendBeaconCloseRequest) => void;

export type ExpoErrorLocation = {
  line: number,
  column: number,
};

export type ExpoPubnubError = {
  device?: ExpoDevice,
  loc?: ExpoErrorLocation,
  message?: string,
  stack?: string,
  line?: number,
  column?: number,
};

export type ExpoError = {
  device?: ExpoDevice,
  startLine?: number,
  endLine?: number,
  startColumn?: number,
  endColumn?: number,
  message: string,
  stack?: string,
};

export type ExpoPubnubDeviceLog = {
  type: 'CONSOLE',
  device: ExpoDevice,
  method: 'log' | 'error' | 'warn',
  payload: Array<any>,
};

// `console.log`, `console.warn`, `console.error`
export type ExpoDeviceLog = {
  device: ExpoDevice,
  method: 'log' | 'warn' | 'error',
  message: string,
  arguments: any, // the raw fields that were passed to the console.* call
};

export type ExpoDevice = {
  name: string,
  id: string,
  platform: string,
};

export type ExpoPresenceStatus = 'join' | 'leave';

export type ExpoPresenceEvent = {
  device: ExpoDevice,
  status: ExpoPresenceStatus,
};

export type ExpoStateEvent = {
  files: ExpoSnackFiles,
  sdkVersion: SDKVersion,
  name: ?string,
  description: ?string,
  dependencies: any, // TODO: more specific
  isSaved: boolean,
  isResolving: boolean,
};

export type ExpoDependencyV1 = { [name: string]: string };

export type ExpoDependencyV2 = {
  [name: string]: {
    version: string, // currently specific version, can expand to semver range, git url, snack url, js file on the web
    // isPeerDep: boolean, // may need to have importing snacks make sense
    resolved?: string, // result of snackager processing the resource
    isUserSpecified: boolean, // can adjust version to resolve peerDeps if false
    peerDependencies?: {
      [name: string]: {
        version: string,
      },
    },
  },
};

export type ExpoDependencyResponse = {
  name: string,
  version: string,
  handle: string,
  dependencies?: { [key: string]: string },
  error?: Error,
};

export type ExpoStatusResponse = {
  type: 'STATUS_REPORT',
  previewLocation: string,
  status: boolean,
};

export type ExpoMessagingListeners = {
  message(payload: {
    message:
      | ExpoPubnubDeviceLog
      | ExpoStatusResponse
      | { type: 'RESEND_CODE', device: ExpoDevice }
      | { type: 'ERROR', error?: string, device: ExpoDevice },
  }): void,

  presence(payload: { action: 'join' | 'leave' | 'timeout', uuid: string }): void,

  status(payload: {
    category:
      | 'PNConnectedCategory'
      | 'PNNetworkDownCategory'
      | 'PNNetworkIssuesCategory'
      | 'PNReconnectedCategory'
      | 'PNNetworkUpCategory',
  }): void,
};

export type Transport = 'PubNub' | 'postMessage';

export interface ExpoMessaging {
  addListener(options: ExpoMessagingListeners): void;

  publish(
    channel: string,
    message:
      | { type: 'RELOAD_SNACK' }
      | { type: 'REQUEST_STATUS' }
      | {
          type: 'LOADING_MESSAGE',
          message: string,
        }
      | {
          type: 'CODE',
          diff: { [key: string]: string },
          s3url: { [key: string]: string },
          dependencies: ExpoDependencyV2,
          metadata: { [key: string]: any },
        },
    transports: Transport[]
  ): Promise<any[]>;

  subscribe(channel: string, transports: Transport[]): void;

  unsubscribe(channel: string, transports: Transport[]): void;
}

export type Platform = 'android' | 'ios' | 'web';
