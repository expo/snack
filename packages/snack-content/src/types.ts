import { SDKVersion, SDKFeature } from './sdks/types';

export type { SDKVersion, SDKFeature };

/**
 * The content of a Snack code or asset file.
 */
export type SnackFile = SnackCodeFile | SnackAssetFile;

/**
 * A non-asset file that is included with the project.
 * This can be either a code file (.js/.tsx) or a support
 * file such as a markdown or JSON file.
 */
export type SnackCodeFile = {
  type: 'CODE';
  contents: string;
  error?: Error;
};

/**
 * An asset file that refers to externally available
 * content such as an image or font.
 *
 * When resolved, the `contents` field is an URL to the
 * uploaded asset. A File, Blob or FormData object may
 * also be provided after which it is automatically uploaded
 * and converted into an URL.
 */
export type SnackAssetFile = {
  type: 'ASSET';
  contents: string | File | Blob | FormData; // string = url
  error?: Error;
};

/**
 * Dictionary of filenames and their content that make up
 * the files of the Snack.
 */
export type SnackFiles = {
  [path: string]: SnackFile;
};

/**
 * Dictionary of dependencies and their version.
 */
export type SnackDependencyVersions = { [name: string]: string };

/**
 * The version, resolved handle, peer-dependencies and optional
 * error of a dependency.
 */
export type SnackDependency = {
  version: string;
  handle?: string;
  peerDependencies?: SnackDependencyVersions;
  error?: Error;
  wantedVersion?: string;
};

/**
 * Dictionary of dependency names and their (resolved) versions.
 */
export type SnackDependencies = {
  [name: string]: SnackDependency;
};

/**
 * Wanted version of the dependency that is missing, including the
 * dependants which have this dependency as a peer-dependency.
 */
export type SnackMissingDependency = {
  dependents: string[];
  wantedVersion?: string;
};

/**
 * Dictionary of dependencies that are missing.
 */
export type SnackMissingDependencies = {
  [name: string]: SnackMissingDependency;
};
