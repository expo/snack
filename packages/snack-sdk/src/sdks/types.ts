/**
 * Version of the sdk to use (e.g. "42.0.0").
 */
export type SDKVersion = '39.0.0' | '40.0.0' | '41.0.0' | '42.0.0' | '43.0.0';

/** @internal */
export type SDKSpec = {
  // Version-spec for the published "expo" package. This version is
  // used to fetch compatible package versions. The value is typically
  // in the form of "^39.0.0" but may also contain custom specs, such
  // as "40.0.0-beta.2".
  version: string;

  // Modules that are pre-loaded by the Snack runtime, and which
  // the user does not need to add to `package.json`.
  // The version for these packages is hardcoded as it cannot be
  // resolved using bundledNativeModules.json.
  coreModules: {
    [name: string]: string;
  };

  // Modules that are pre-loaded by the Snack runtime, but which
  // are expected to be added as a dependency anyway. This ensures
  // that `package.json` is up to date, when downloading the Snack
  // as a zip-file. These modules don't have versions defined, but
  // instead bundledNativeModules.json should be used to obtain
  // the recent version.
  bundledModules: {
    [name: string]: '*';
  };

  // Modules that have been deprecated. The value represents a valid
  // a description providing a solution or explanation (eg.
  // "expo-permissions": "Use permissions getters and requesters in specific modules instead...""
  deprecatedModules?: {
    [name: string]: string;
  };
};

/**
 * Feature that is supported by the SDK (e.g. "TYPESCRIPT").
 */
export type SDKFeature =
  | 'MULTIPLE_FILES'
  | 'PROJECT_DEPENDENCIES'
  | 'TYPESCRIPT'
  | 'UNIMODULE_IMPORTS'
  | 'POSTMESSAGE_TRANSPORT'
  | 'VERSIONED_SNACKAGER';
