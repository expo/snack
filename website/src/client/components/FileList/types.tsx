export type TextFileEntry = Readonly<{
  item: {
    path: string;
    type: 'file';
    content: string;
    virtual?: true;
    asset?: false;
  };
  state: {
    isOpen?: boolean;
    isFocused?: boolean;
    isSelected?: boolean;
    isCreating?: boolean;
    isExpanded?: false;
    isError?: boolean;
    isLoading?: boolean;
  };
}>;

export type AssetFileEntry = Readonly<{
  item: {
    path: string;
    type: 'file';
    uri: string;
    asset: true;
    virtual?: true;
  };
  state: {
    isOpen?: boolean;
    isFocused?: boolean;
    isSelected?: boolean;
    isCreating?: boolean;
    isExpanded?: false;
    isError?: boolean;
    isLoading?: boolean;
  };
}>;

export type FolderEntry = Readonly<{
  item: {
    path: string;
    type: 'folder';
    asset?: false;
    virtual?: false;
  };
  state: {
    isOpen?: boolean;
    isFocused?: boolean;
    isExpanded?: boolean;
    isSelected?: boolean;
    isCreating?: boolean;
    isError?: boolean;
    isLoading?: boolean;
  };
}>;

export type FileSystemEntry = TextFileEntry | AssetFileEntry | FolderEntry;
