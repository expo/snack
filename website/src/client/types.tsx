import * as React from 'react';
import {
  SnackState,
  SnackFile,
  SnackCodeFile,
  SnackAssetFile,
  SnackFiles,
  SnackDependencies,
  SnackDependency,
  SnackMissingDependencies,
  SnackConnectedClient,
  SnackConnectedClients,
  SDKVersion,
} from 'snack-sdk';

import { ThemeName } from './components/Preferences/withThemeName';

export type {
  SnackState,
  SnackFile,
  SnackCodeFile,
  SnackAssetFile,
  SnackFiles,
  SnackDependencies,
  SnackDependency,
  SnackMissingDependencies,
  SnackConnectedClient,
  SnackConnectedClients,
  SDKVersion,
  ThemeName,
};

export type SnackManifest = {
  name: string;
  description: string;
  sdkVersion?: SDKVersion;
};

export type SavedSnack = {
  id: string;
  created: string;
  code?: string | SnackFiles;
  manifest: SnackManifest;
  dependencies?: SnackDependencies;
  history?: SaveHistory;
  isDraft?: boolean;
};

export type SnackDefaults = {
  name: string;
  channel: string;
};

export type RouterData =
  | {
      type: 'success';
      snack?: SavedSnack;
      defaults: SnackDefaults;
    }
  | {
      type: 'error';
      error: { message: string };
      defaults: SnackDefaults;
    };

export type Viewer = {
  username: string;
  nickname: string;
  picture?: string;
  user_metadata?: {
    appetize_code: string;
  };
};

export type Platform = 'android' | 'ios' | 'web' | 'mydevice';

export type Device = {
  name: string;
  id: string;
  platform: Platform;
  status: 'connected' | 'disconnected';
  timestamp: number;
};

export type DeviceLog = {
  device: Device;
  method: 'log' | 'error' | 'warn';
  payload: unknown[];
};

export type QueryInitParams = {
  code?: string;
  sourceUrl?: string;
  name?: string;
  description?: string;
  dependencies?: string;
  files?: string;
  sdkVersion?: SDKVersion;
  iframeId?: string;
  waitForData?: 'boolean';
  saveToAccount?: 'true' | 'false';
};

export type QueryStateParams = {
  preview?: 'true' | 'false';
  platform?: Platform;
  theme?: ThemeName;
  supportedPlatforms?: string;
  appetizePayerCode?: string;
  verbose?: 'true' | 'false';
  hideQueryParams?: 'true' | 'false';
};

export type QueryParams = QueryInitParams & QueryStateParams;

export type SaveStatus =
  | 'unsaved'
  | 'edited'
  | 'saving-draft'
  | 'saved-draft'
  | 'publishing'
  | 'published';

export type SaveHistory = {
  hashId: string;
  savedAt: string;
  isDraft?: boolean;
}[];

export type SaveOptions = {
  isDraft?: boolean;
  ignoreUser?: boolean;
  excludeFromHistory?: boolean;
};

export enum AnnotationSeverity {
  LOADING = -1,
  IGNORE = 0,
  INFO = 1,
  WARNING = 2,
  ERROR = 3,
  FATAL = 4,
}

export type AnnotationAction = {
  title: string;
  icon?: React.ComponentType<any>;
  run: () => void;
};

export type AnnotationLocation = {
  fileName: string;
  startLineNumber: number;
  endLineNumber: number;
  startColumn: number;
  endColumn: number;
};

export type Annotation = {
  message: string;
  severity: AnnotationSeverity;
  source: 'Device' | 'Web' | 'JSON' | 'ESLint' | 'Dependencies';
  location?: AnnotationLocation;
  action?: AnnotationAction;
};

export type $SetComplement<A, A1 extends A> = A extends A1 ? never : A;

export type $Subtract<T extends T1, T1 extends object> = Pick<T, $SetComplement<keyof T, keyof T1>>;
