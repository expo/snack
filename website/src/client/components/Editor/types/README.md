# Vendored types

Some libraries are bundled inside the Runtime, without their types. This means we have to manually add them to Monaco.

## Existing types

- `@shopify/react-native-skia/dist/web` â†’ See `runtime/src/NativeModules/ReactNativeSkia.tsx`.
