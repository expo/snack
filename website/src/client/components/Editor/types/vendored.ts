/** All hard-coded and vendored types */
export const vendoredTypes: Record<string, string> = {
  // Workaround for React 17+ and auto jsx runtime
  // See: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d7b179c7a9a4aa4ff13f0608606ae10a94349014/types/react/jsx-runtime.d.ts#L2
  ...makeModuleType('react/jsx-runtime', `import 'react';`),
  // See: /runtime/src/NativeModules/ReactNativeSkia.tsx
  ...makeModuleType(
    '@shopify/react-native-skia/dist/web',
    `declare module "@shopify/react-native-skia/dist/web" {
      import { Suspense, ComponentProps, ComponentType } from 'react';
      interface WithSkiaProps {
          fallback?: ComponentProps<typeof Suspense>['fallback'];
          getComponent: () => Promise<{
              default: ComponentType;
          }>;
      }
      export function WithSkia({ fallback, getComponent }: WithSkiaProps): JSX.Element;
      export function LoadSkia(): Promise<void>;
    }`
  ),
};

/**
 * Create a declaration file for a specific import path.
 * This outputs both the `<pkg>/package.json` and `<pkg>/index.d.ts` files.
 * The package file needs to point to the loaded declaration file, without that the types are not picked up.
 */
function makeModuleType(importName: string, declarationContent: string): Record<string, string> {
  return {
    // For each import path, we need to fake a package file, pointing to the right declarations.
    [`node_modules/${importName}/package.json`]: JSON.stringify({
      name: importName,
      version: '1.0.0',
      types: './index.d.ts',
    }),
    // The declaration also needs to be wrapped inside `declare module "<pkg>" { ... }`
    [`node_modules/${importName}/index.d.ts`]: declarationContent,
  };
}
