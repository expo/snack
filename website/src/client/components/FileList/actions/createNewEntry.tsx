import { isInsideFolder } from '../../../utils/fileUtilities';
import { FileSystemEntry } from '../types';
import createEntryAtPath from './createEntryAtPath';
import updateEntry from './updateEntry';

export default function createNewEntry(
  entries: FileSystemEntry[],
  type: 'file' | 'folder',
  at?: string | undefined
) {
  const e: FileSystemEntry =
    type === 'file'
      ? {
          item: {
            path: 'Untitled file.js',
            type: 'file',
            content: '',
          },
          state: {
            isCreating: true,
          },
        }
      : {
          item: {
            path: 'Untitled folder',
            type: 'folder',
          },
          state: {
            isCreating: true,
          },
        };

  let path = at;

  if (typeof path !== 'string') {
    // Get the current folder if no path is specified
    const entry = entries.find((e) => e.state.isSelected === true);

    path = entry ? entry.item.path : undefined;
  }

  const entry = createEntryAtPath(entries, path, e);

  const next = entries.map((e) => {
    // Expand the parent folder
    if (
      e.item.type === 'folder' &&
      isInsideFolder(entry.item.path, e.item.path) &&
      !e.state.isExpanded
    ) {
      return updateEntry(e, {
        state: { isExpanded: true },
      });
    }

    return e;
  });

  return [...next, entry];
}
