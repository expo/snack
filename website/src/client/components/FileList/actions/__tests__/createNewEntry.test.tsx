import { FileSystemEntry } from '../../types';
import createNewEntry from '../createNewEntry';

it('creates new file', () => {
  const entries: FileSystemEntry[] = [
    { item: { path: 'test', type: 'folder' }, state: {} },
    { item: { path: 'test/App.js', type: 'file', content: '' }, state: {} },
    { item: { path: 'components', type: 'folder' }, state: {} },
  ];

  expect(createNewEntry(entries, 'file')).toMatchSnapshot();

  expect(createNewEntry(entries, 'file', 'test')).toMatchSnapshot();
});

it('creates new folder', () => {
  const entries: FileSystemEntry[] = [
    { item: { path: 'test', type: 'folder' }, state: {} },
    { item: { path: 'test/App.js', type: 'file', content: '' }, state: {} },
    { item: { path: 'components', type: 'folder' }, state: {} },
  ];

  expect(createNewEntry(entries, 'folder')).toMatchSnapshot();

  expect(createNewEntry(entries, 'folder', 'test')).toMatchSnapshot();
});
