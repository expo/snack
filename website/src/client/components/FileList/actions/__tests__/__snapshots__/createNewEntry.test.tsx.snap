// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates new file 1`] = `
Array [
  Object {
    "item": Object {
      "path": "test",
      "type": "folder",
    },
    "state": Object {},
  },
  Object {
    "item": Object {
      "content": "",
      "path": "test/App.js",
      "type": "file",
    },
    "state": Object {},
  },
  Object {
    "item": Object {
      "path": "components",
      "type": "folder",
    },
    "state": Object {},
  },
  Object {
    "item": Object {
      "content": "",
      "path": "Untitled file.js",
      "type": "file",
    },
    "state": Object {
      "isCreating": true,
    },
  },
]
`;

exports[`creates new file 2`] = `
Array [
  Object {
    "item": Object {
      "path": "test",
      "type": "folder",
    },
    "state": Object {
      "isExpanded": true,
    },
  },
  Object {
    "item": Object {
      "content": "",
      "path": "test/App.js",
      "type": "file",
    },
    "state": Object {},
  },
  Object {
    "item": Object {
      "path": "components",
      "type": "folder",
    },
    "state": Object {},
  },
  Object {
    "item": Object {
      "content": "",
      "path": "test/Untitled file.js",
      "type": "file",
    },
    "state": Object {
      "isCreating": true,
    },
  },
]
`;

exports[`creates new folder 1`] = `
Array [
  Object {
    "item": Object {
      "path": "test",
      "type": "folder",
    },
    "state": Object {},
  },
  Object {
    "item": Object {
      "content": "",
      "path": "test/App.js",
      "type": "file",
    },
    "state": Object {},
  },
  Object {
    "item": Object {
      "path": "components",
      "type": "folder",
    },
    "state": Object {},
  },
  Object {
    "item": Object {
      "path": "Untitled folder",
      "type": "folder",
    },
    "state": Object {
      "isCreating": true,
    },
  },
]
`;

exports[`creates new folder 2`] = `
Array [
  Object {
    "item": Object {
      "path": "test",
      "type": "folder",
    },
    "state": Object {
      "isExpanded": true,
    },
  },
  Object {
    "item": Object {
      "content": "",
      "path": "test/App.js",
      "type": "file",
    },
    "state": Object {},
  },
  Object {
    "item": Object {
      "path": "components",
      "type": "folder",
    },
    "state": Object {},
  },
  Object {
    "item": Object {
      "path": "test/Untitled folder",
      "type": "folder",
    },
    "state": Object {
      "isCreating": true,
    },
  },
]
`;
